Src/Web/Web.Shared.Net/ApplicationInsightsHttpModule.cs

```diff
@@ -121,29 +145,15 @@ private void AddCorreleationHeaderOnSendRequestHeaders(HttpApplication httpAppli
                 {
                     // We use reflection here because 'AddOnSendingHeaders' is only available post .net framework 4.5.2. Hence we call it if we can find it.
                     // Not using reflection would result in MissingMethodException when 4.5 or 4.5.1 is present. 
-                    MethodInfo addOnSendingHeadersMethod = httpApplication.Response.GetType().GetMethod("AddOnSendingHeaders");
-
-                    if (addOnSendingHeadersMethod != null)
+                    if (this.addOnSendingHeadersMethodExists == null)
                     {
-                        var parameters = new object[]
-                        {
-                            new Action<HttpContext>((httpContext) =>
-                            {
-                                try
-                                {
-                                    if (this.requestModule != null)
-                                    {
-                                        this.requestModule.AddTargetHashForResponseHeader(httpApplication.Context);
-                                    }
-                                }
-                                catch (Exception ex)
-                                {
-                                    WebEventSource.Log.AddTargetHeaderFailedWarning(ex.ToInvariantString());
-                                }
-                            })
-                        };
-
-                        addOnSendingHeadersMethod.Invoke(httpApplication.Response, parameters);
+                        this.addOnSendingHeadersMethod = httpApplication.Response.GetType().GetMethod("AddOnSendingHeaders");
+                        this.addOnSendingHeadersMethodExists = this.addOnSendingHeadersMethod != null;
+                    }
+                    
+                    if (this.addOnSendingHeadersMethod != null)
+                    {                        
+                        this.addOnSendingHeadersMethod.Invoke(httpApplication.Response, this.paramsForReflectiveCall);
```

Comment:
Invoke is more expensive than delegate. See https://gist.github.com/jmangelo/3173109#file-gistfile1-cs-L57-L66 for example use